src/bytes.rs

```diff
@@ -418,7 +418,15 @@ impl Bytes {
     #[inline]
     pub fn truncate(&mut self, len: usize) {
         if len < self.len {
-            self.len = len;
+            // The Vec "promotable" vtables do not store the capacity,
+            // so we cannot truncate while using this repr. We *have* to
+            // promote using `split_off` so the capacity can be stored.
+            if self.vtable as *const Vtable == &PROMOTABLE_EVEN_VTABLE ||
+                self.vtable as *const Vtable == &PROMOTABLE_ODD_VTABLE {
+                drop(self.split_off(len));
```

Comment:
Might want to consider implementing this by using `clone` directly?
```rust
*self = self.clone();
self.len = len;
```
Which should accomplish the upgrade in-place (and update the local vtable to boot), while avoiding any extra work associated with `split_off`.

Reply:
`clone()` will be less efficient at `truncate(0)`.